### PR TITLE
HADOOP-16635. S3A innerGetFileStatus scans for directories-only still does a HEAD.

### PR DESCRIPTION
**Pull request by steveloughran**
1. Make sure the probe set is passed down.
2. check for Head probe wraps only HEAD request.

Change-Id: I7664a80da53f85c42c319cd6b56b3d6366e09fc6

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 38 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1148 | trunk passed |
| +1 | compile | 36 | trunk passed |
| +1 | checkstyle | 28 | trunk passed |
| +1 | mvnsite | 41 | trunk passed |
| +1 | shadedclient | 803 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | trunk passed |
| 0 | spotbugs | 62 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 60 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 35 | the patch passed |
| +1 | compile | 28 | the patch passed |
| +1 | javac | 28 | the patch passed |
| +1 | checkstyle | 19 | the patch passed |
| +1 | mvnsite | 33 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 825 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 24 | the patch passed |
| +1 | findbugs | 63 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 90 | hadoop-aws in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 3425 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1601 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 6d696d8ccfbf 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 8de4374 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/1/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 40 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1061 | trunk passed |
| +1 | compile | 36 | trunk passed |
| +1 | checkstyle | 27 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 811 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | trunk passed |
| 0 | spotbugs | 60 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 29 | the patch passed |
| +1 | javac | 29 | the patch passed |
| +1 | checkstyle | 21 | the patch passed |
| +1 | mvnsite | 33 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 789 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 63 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 89 | hadoop-aws in the patch passed. |
| +1 | asflicense | 34 | The patch does not generate ASF License warnings. |
| | | 3322 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1601 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 18d2feefbc92 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b8086bf |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/2/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 41 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1070 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 28 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 785 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | trunk passed |
| 0 | spotbugs | 60 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 59 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 29 | the patch passed |
| +1 | javac | 29 | the patch passed |
| +1 | checkstyle | 20 | the patch passed |
| +1 | mvnsite | 33 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 780 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 62 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 90 | hadoop-aws in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 3296 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1601 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 18c4809af70f 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1a77a15 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/3/testReport/ |
| Max. process+thread count | 447 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by sidseth** [comment]
Not really an expert on the S3AFs core functionality. That said, the current code does seem quite broken - if HEAD is not included in the probe set.

```
if (!key.endsWith("/") && probes.contains(StatusProbeEnum.DirMarker)) {
```
This may need a minor change. I believe the leading key.endsWith("/") is to avoid duplicating a HEAD request - if the key already contains a "/". However, if the set of probes does not contain "Head", and it does contain "DirMarker", but the key ends in a "/" - there won't be any HEAD requests. Is that expected?

Other than this, looks good to me.

General question, unrelated to the patch, is a single LIST not sufficient to cover both the with and without "/" scenario?

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 42 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1063 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 27 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 800 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | trunk passed |
| 0 | spotbugs | 59 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 58 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 29 | the patch passed |
| +1 | javac | 29 | the patch passed |
| +1 | checkstyle | 21 | the patch passed |
| +1 | mvnsite | 33 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 774 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 61 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 80 | hadoop-aws in the patch passed. |
| +1 | asflicense | 34 | The patch does not generate ASF License warnings. |
| | | 3292 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1601 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux c0c2bf00b7f4 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 6f1ab95 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/4/testReport/ |
| Max. process+thread count | 449 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
Sid, thanks for the comments, will review/update the patch

Interesting point about the double list. This code path is how its always been, presumably descended from the s3n code. LIST is slower, costs more and much more prone to eventual consistency, which are all good arguments for HEAD first.

I actually plan to tune some of the calls which always seem to get used on directory walks (listStatus, listFiles, listLocatedStatus) to do the subtree list first, and only go for the HEAD calls if they don't find any children. This is to reduce the cost of treewalks where the bias is towards populated directories

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java [hunk 1]
```diff
@@ -2730,39 +2730,41 @@ S3AFileStatus innerGetFileStatus(final Path f,
...
-      try {
-        ObjectMetadata meta = getObjectMetadata(key);
-
-        if (objectRepresentsDirectory(key, meta.getContentLength())) {
-          LOG.debug("Found exact file: fake directory");
-          return new S3AFileStatus(Tristate.TRUE, path, username);
-        } else {
-          LOG.debug("Found exact file: normal file");
+    if (!key.isEmpty()) {
+      if (probes.contains(StatusProbeEnum.Head) && !key.endsWith("/")) {
```
Comment (sidseth) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: Not sure if this is ready to review.
With this change - if someone asks for a "HEAD" only, on a URL which happens to have a trailing "/" nothing will be looked up.
IF HAED + DIRMARKER - then yes, at least 1 request will go out.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 49 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1149 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 27 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 819 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | trunk passed |
| 0 | spotbugs | 61 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 58 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 36 | the patch passed |
| +1 | compile | 30 | the patch passed |
| +1 | javac | 30 | the patch passed |
| +1 | checkstyle | 20 | the patch passed |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 781 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | the patch passed |
| +1 | findbugs | 65 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 86 | hadoop-aws in the patch passed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 3400 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1601 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 8443bcc90283 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9c72bf4 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/5/testReport/ |
| Max. process+thread count | 443 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java (hunk 1 above)
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: yes. That's exactly my thought. 
note: none of this API is public, its for avoiding problems on ops where we don't want to look for a file but do for a dir marker. And AFAIK, you can't go from a Path to a / as we strip that off.

How about
1.  I do a review of all places where we don't ask for the HEAD? So far I think I'm only doing it in create
1. I clarify in javadocs

**Comment by steveloughran** [comment]
updated the docs. The only place we don't do Head and dir marker is in create()

Now. can you create a Path with a trailing / ? I was about to say no, but remembered https://issues.apache.org/jira/browse/HADOOP-15430 .. one of the constructors of Path does let you get away with it, which is something which breaks S3Guard already

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 82 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1280 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 24 | trunk passed |
| +1 | mvnsite | 38 | trunk passed |
| +1 | shadedclient | 883 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | trunk passed |
| 0 | spotbugs | 63 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 61 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 28 | the patch passed |
| +1 | javac | 28 | the patch passed |
| +1 | checkstyle | 19 | the patch passed |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 876 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 22 | the patch passed |
| +1 | findbugs | 64 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 69 | hadoop-aws in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 3698 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1601 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 22c4d6fb1156 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ec86f42 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/6/testReport/ |
| Max. process+thread count | 424 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1601/6/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java (hunk 1 above)
Comment (bgaborg) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: this is handled in https://issues.apache.org/jira/browse/HADOOP-15430 right?

**Comment by steveloughran** [comment]
thx
-merged

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java (hunk 1 above)
Comment (bgaborg) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: @steveloughran

**Pull request closed without merging**